### PR TITLE
Operator SDK decommissioning  

### PR DESCRIPTION
**Pull request by mjudeikis**
### Which issue this PR addresses:

Fixes https://msazure.visualstudio.com/AzureRedHatOpenShift/_workitems/edit/9534793

### What this PR does / why we need it:

Each time we need to touch Operator SDK code in our repo little cat in the world dies from stress. 

### Test plan for issue:

Unit tests

### Is there any documentation that needs to be updated for this PR?

No

### CODE REVIEW

**Review by troy0820 — COMMENTED** [review]
LGTM

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (petrkotas) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:99]: The code here is duplicating the [`corev1` API](https://github.com/Azure/ARO-RP/blob/master/vendor/k8s.io/api/core/v1/types.go#L2393) or operator related https://github.com/Azure/ARO-RP/blob/d8d2303a9a7f1074ff87dde9d04d701cd0e3ef69/vendor/github.com/openshift/api/operator/v1/types.go#L166

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (petrkotas) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:105]: Another duplicate 

https://github.com/Azure/ARO-RP/blob/master/vendor/k8s.io/api/core/v1/types.go#L2399-L2403

https://github.com/Azure/ARO-RP/blob/d8d2303a9a7f1074ff87dde9d04d701cd0e3ef69/vendor/github.com/openshift/api/operator/v1/types.go#L168-L172

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (petrkotas) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:112]: Here https://github.com/Azure/ARO-RP/blob/master/vendor/k8s.io/api/core/v1/types.go#L2520

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (petrkotas) [inline pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:129]: and here https://github.com/Azure/ARO-RP/blob/master/vendor/k8s.io/api/core/v1/types.go#L2543-L2563

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (petrkotas) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go:112]: I understand this is for the operator, but it is running in a pod, so it is a podcondition.

**Review by petrkotas — CHANGES_REQUESTED** [review]
I am happy we removed the operator-framework, thank you for that.
I have just a simple ask, would you please consider reducing the amount of duplicated code added by this PR? Similar or same code is either in the vendored openshift API or vendored kubernetes API.

Thank you

**Comment by mjudeikis** [comment]
Very good catch :) checked apimachinery but not core! :D

**Review by petrkotas — DISMISSED** [review]
Perfect. Thank you.

**Review by nilsanderselde — DISMISSED** [review]
Looks good

File: test/e2e/operator.go
Comment (m1kola) [inline test/e2e/operator.go:332]:
```suggestion
```

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:69]: What "with condition change"?

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:38]: This might benefit fro better naming. Why noop?

Comment (m1kola) [reply pkg/util/conditions/condition_test.go:38]: This actually triggers an update because of:

```go
	if role == operator.RoleMaster && cluster.Status.OperatorVersion != version.GitCommit {
		cluster.Status.OperatorVersion = version.GitCommit
		changed = true
	}
```

Which highlights that `SetCondition` name is confusing & that it probably doing more than it should.

**Review by m1kola — CHANGES_REQUESTED** [review]
Few suggestions/comments.

File: pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
Comment (mjudeikis) [reply pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go]: You are right. This is what you get when you do multiple things at the same time :D

File: pkg/util/conditions/condition_test.go
Comment (mjudeikis) [reply pkg/util/conditions/condition_test.go:39]: yap. It treats version as condition. 
Not much we can do here (not the place too). 
Added test to over this for now

File: pkg/util/conditions/condition_test.go
Comment (mjudeikis) [reply pkg/util/conditions/condition_test.go]: updated

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:75]: What are we testing in this case? That the version is being updated? If so, then probably we should reflect it in the case name

```suggestion
		{
			name: "update operator version",
			aroclient: arofake.NewSimpleClientset(&arov1alpha1.Cluster{
				ObjectMeta: metav1.ObjectMeta{
					Name: objectName,
				},
				Status: arov1alpha1.ClusterStatus{
					Conditions: []operatorv1.OperatorCondition{
						{
							Type:   arov1alpha1.InternetReachableFromMaster.String(),
							Status: operatorv1.ConditionFalse,
						},
					},
					OperatorVersion: "?",
				},
			}),
			expected: arov1alpha1.ClusterStatus{
				Conditions: []operatorv1.OperatorCondition{
					{
						Type:   arov1alpha1.InternetReachableFromMaster.String(),
						Status: operatorv1.ConditionFalse,
					},
				},
				OperatorVersion: version,
			},
		},
```

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:100]: What are we testing in this case? That the operator version is being update from empty to the current version? If so, then I think we can remove this test case as the previous one covers it already.

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:130]: I think this test case can be removed in favour of the next one.

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:132]: Preserve what? Maybe something like this reflects the test case better?

```suggestion
			name: "update one of the existing conditions",
```

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:21]: I think we need the following test cases:

* Case where we exit early and do not call `UpdateStatus` (is `noop` case supposed to test this? See [this thread](https://github.com/Azure/ARO-RP/pull/1508/files#r656952649))
* Case where we set a new condition (current conditions slice is empty, but expected result would be a new condition from the `input`).
* Case where an existing condition gets updated with the condition which is exactly the same: we want to make sure that transition time doesn't change

File: pkg/util/conditions/condition_test.go
Comment (m1kola) [inline pkg/util/conditions/condition_test.go:33]: Make it pointer maybe?

File: pkg/util/conditions/condition.go
Comment (m1kola) [inline pkg/util/conditions/condition.go:32]: Potential nil pointer dereference

**Review by m1kola — CHANGES_REQUESTED** [review]
I think this is very close. I only spoted one potential nil pointer dereference + left several commetns about tests.

**Comment by mjudeikis** [comment]
Either lets merge it or drop it. This so much nit'ing I don't think it is worth time stent now.

**Comment by m1kola** [comment]
@mjudeikis some of the comments were nits, I agree. But tests were not covering several cases, but at the same time had duplicates. 

I know you are busy, so I prepared a patch for the remainig feedback.

```patch
diff --git a/pkg/monitor/cluster/arooperatorconditions.go b/pkg/monitor/cluster/arooperatorconditions.go
index f090d4e6..1c0b6269 100644
--- a/pkg/monitor/cluster/arooperatorconditions.go
+++ b/pkg/monitor/cluster/arooperatorconditions.go
@@ -13,7 +13,7 @@ import (
 	arov1alpha1 "github.com/Azure/ARO-RP/pkg/operator/apis/aro.openshift.io/v1alpha1"
 )
 
-var aroOperatorConditionsExpected = map[arov1alpha1.CondititionType]operatorv1.ConditionStatus{
+var aroOperatorConditionsExpected = map[string]operatorv1.ConditionStatus{
 	arov1alpha1.InternetReachableFromMaster: operatorv1.ConditionTrue,
 	arov1alpha1.InternetReachableFromWorker: operatorv1.ConditionTrue,
 }
@@ -25,7 +25,7 @@ func (mon *Monitor) emitAroOperatorConditions(ctx context.Context) error {
 	}
 
 	for _, c := range cluster.Status.Conditions {
-		if aroOperatorConditionsExpected[arov1alpha1.CondititionType(c.Type)] == c.Status {
+		if aroOperatorConditionsExpected[c.Type] == c.Status {
 			continue
 		}
 
diff --git a/pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go b/pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
index d78bdac6..72cd9865 100644
--- a/pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
+++ b/pkg/operator/apis/aro.openshift.io/v1alpha1/cluster_types.go
@@ -16,17 +16,15 @@ const (
 	ServicePrincipalValid       = "ServicePrincipalValid"
 )
 
-type CondititionType string
-
 // AllConditionTypes is a operator conditions currently in use, any condition not in this list is not
 // added to the operator.status.conditions list
-func AllConditionTypes() []CondititionType {
-	return []CondititionType{InternetReachableFromMaster, InternetReachableFromWorker, MachineValid, ServicePrincipalValid}
+func AllConditionTypes() []string {
+	return []string{InternetReachableFromMaster, InternetReachableFromWorker, MachineValid, ServicePrincipalValid}
 }
 
 // ClusterChecksTypes represents checks performed on the cluster to verify basic functionality
-func ClusterChecksTypes() []CondititionType {
-	return []CondititionType{InternetReachableFromMaster, InternetReachableFromWorker, MachineValid, ServicePrincipalValid}
+func ClusterChecksTypes() []string {
+	return []string{InternetReachableFromMaster, InternetReachableFromWorker, MachineValid, ServicePrincipalValid}
 }
 
 type GenevaLoggingSpec struct {
diff --git a/pkg/util/conditions/condition.go b/pkg/util/conditions/condition.go
index afedb121..e1e5531d 100644
--- a/pkg/util/conditions/condition.go
+++ b/pkg/util/conditions/condition.go
@@ -47,15 +47,15 @@ func SetCondition(ctx context.Context, arocli aroclient.Interface, cond *operato
 	})
 }
 
-func IsTrue(conditions []operatorv1.OperatorCondition, t arov1alpha1.CondititionType) bool {
+func IsTrue(conditions []operatorv1.OperatorCondition, t string) bool {
 	return isCondition(conditions, t, operatorv1.ConditionTrue)
 }
 
-func IsFalse(conditions []operatorv1.OperatorCondition, t arov1alpha1.CondititionType) bool {
+func IsFalse(conditions []operatorv1.OperatorCondition, t string) bool {
 	return isCondition(conditions, t, operatorv1.ConditionFalse)
 }
 
-func isCondition(conditions []operatorv1.OperatorCondition, t arov1alpha1.CondititionType, s operatorv1.ConditionStatus) bool {
+func isCondition(conditions []operatorv1.OperatorCondition, t string, s operatorv1.ConditionStatus) bool {
 	for _, condition := range conditions {
 		if condition.Type == string(t) && condition.Status == s {
 			return true
@@ -70,13 +70,13 @@ func cleanStaleConditions(cluster *arov1alpha1.Cluster, role string) (changed bo
 	conditions := make([]operatorv1.OperatorCondition, 0, len(cluster.Status.Conditions))
 
 	// cleanup any old conditions
-	current := map[arov1alpha1.CondititionType]bool{}
+	current := map[string]bool{}
 	for _, ct := range arov1alpha1.AllConditionTypes() {
 		current[ct] = true
 	}
 
 	for _, cond := range cluster.Status.Conditions {
-		if _, ok := current[arov1alpha1.CondititionType(cond.Type)]; ok {
+		if _, ok := current[cond.Type]; ok {
 			conditions = append(conditions, cond)
 		} else {
 			changed = true
diff --git a/pkg/util/conditions/condition_test.go b/pkg/util/conditions/condition_test.go
index 4ea46562..bcfb4cee 100644
--- a/pkg/util/conditions/condition_test.go
+++ b/pkg/util/conditions/condition_test.go
@@ -7,6 +7,7 @@ import (
 	"context"
 	"reflect"
 	"testing"
+	"time"
 
 	operatorv1 "github.com/openshift/api/operator/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
@@ -32,25 +33,47 @@ func TestSetCondition(t *testing.T) {
 		name      string
 		aroclient aroclient.Interface
 		objects   []runtime.Object
-		input     operatorv1.OperatorCondition
+		input     *operatorv1.OperatorCondition
 
 		expected arov1alpha1.ClusterStatus
 		wantErr  error
 	}{
 		{
-			name: "no condition change",
+			name: "no condition provided",
 			objects: []runtime.Object{&arov1alpha1.Cluster{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: objectName,
 				},
 			}},
+			expected: arov1alpha1.ClusterStatus{},
+		},
+		{
+			name: "new condition provided",
+			objects: []runtime.Object{&arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: objectName,
+				},
+				Status: arov1alpha1.ClusterStatus{
+					Conditions:      []operatorv1.OperatorCondition{},
+					OperatorVersion: version,
+				},
+			}},
+			input: &operatorv1.OperatorCondition{
+				Type:   arov1alpha1.InternetReachableFromMaster,
+				Status: operatorv1.ConditionFalse,
+			},
 			expected: arov1alpha1.ClusterStatus{
+				Conditions: []operatorv1.OperatorCondition{
+					{
+						Type:   arov1alpha1.InternetReachableFromMaster,
+						Status: operatorv1.ConditionFalse,
+					},
+				},
 				OperatorVersion: version,
-				Conditions:      []operatorv1.OperatorCondition{},
 			},
 		},
 		{
-			name: "update operator version",
+			name: "condition provided without status change - only update operator version",
 			objects: []runtime.Object{&arov1alpha1.Cluster{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: objectName,
@@ -65,6 +88,10 @@ func TestSetCondition(t *testing.T) {
 					OperatorVersion: "?",
 				},
 			}},
+			input: &operatorv1.OperatorCondition{
+				Type:   arov1alpha1.InternetReachableFromMaster,
+				Status: operatorv1.ConditionFalse,
+			},
 			expected: arov1alpha1.ClusterStatus{
 				Conditions: []operatorv1.OperatorCondition{
 					{
@@ -76,7 +103,7 @@ func TestSetCondition(t *testing.T) {
 			},
 		},
 		{
-			name: "change with condition",
+			name: "condition provided without status change - no update",
 			objects: []runtime.Object{&arov1alpha1.Cluster{
 				ObjectMeta: metav1.ObjectMeta{
 					Name: objectName,
@@ -84,22 +111,25 @@ func TestSetCondition(t *testing.T) {
 				Status: arov1alpha1.ClusterStatus{
 					Conditions: []operatorv1.OperatorCondition{
 						{
-							Type:   arov1alpha1.InternetReachableFromMaster,
-							Status: operatorv1.ConditionFalse,
+							Type:               arov1alpha1.InternetReachableFromMaster,
+							Status:             operatorv1.ConditionFalse,
+							LastTransitionTime: metav1.Time{Time: time.Date(1970, 0, 0, 0, 0, 0, 0, time.UTC)},
 						},
 					},
+					OperatorVersion: version,
 				},
 			}},
-			input: operatorv1.OperatorCondition{
-				Type:   arov1alpha1.InternetReachableFromMaster,
-				Status: operatorv1.ConditionTrue,
+			input: &operatorv1.OperatorCondition{
+				Type:               arov1alpha1.InternetReachableFromMaster,
+				Status:             operatorv1.ConditionFalse,
+				LastTransitionTime: metav1.Time{Time: time.Date(2021, 0, 0, 0, 0, 0, 0, time.UTC)},
 			},
 			expected: arov1alpha1.ClusterStatus{
 				Conditions: []operatorv1.OperatorCondition{
 					{
 						Type:               arov1alpha1.InternetReachableFromMaster,
-						Status:             operatorv1.ConditionTrue,
-						LastTransitionTime: transitionTime,
+						Status:             operatorv1.ConditionFalse,
+						LastTransitionTime: metav1.Time{Time: time.Date(1970, 0, 0, 0, 0, 0, 0, time.UTC)},
 					},
 				},
 				OperatorVersion: version,
@@ -122,9 +152,10 @@ func TestSetCondition(t *testing.T) {
 							Status: operatorv1.ConditionFalse,
 						},
 					},
+					OperatorVersion: "?",
 				},
 			}},
-			input: operatorv1.OperatorCondition{
+			input: &operatorv1.OperatorCondition{
 				Type:   arov1alpha1.InternetReachableFromMaster,
 				Status: operatorv1.ConditionTrue,
 			},
@@ -143,12 +174,42 @@ func TestSetCondition(t *testing.T) {
 				OperatorVersion: version,
 			},
 		},
+		{
+			name: "cleanup stale conditions",
+			objects: []runtime.Object{&arov1alpha1.Cluster{
+				ObjectMeta: metav1.ObjectMeta{
+					Name: objectName,
+				},
+				Status: arov1alpha1.ClusterStatus{
+					Conditions: []operatorv1.OperatorCondition{
+						{
+							Type:   "staleType",
+							Status: operatorv1.ConditionTrue,
+						},
+					},
+					OperatorVersion: version,
+				},
+			}},
+			input: &operatorv1.OperatorCondition{
+				Type:   arov1alpha1.InternetReachableFromMaster,
+				Status: operatorv1.ConditionTrue,
+			},
+			expected: arov1alpha1.ClusterStatus{
+				Conditions: []operatorv1.OperatorCondition{
+					{
+						Type:               arov1alpha1.InternetReachableFromMaster,
+						Status:             operatorv1.ConditionTrue,
+						LastTransitionTime: transitionTime,
+					},
+				},
+				OperatorVersion: version,
+			},
+		},
 	} {
 		t.Run(tt.name, func(t *testing.T) {
-
 			client := arofake.NewSimpleClientset(tt.objects...)
 
-			err := SetCondition(ctx, client, &tt.input, role)
+			err := SetCondition(ctx, client, tt.input, role)
 			if err != nil && tt.wantErr != nil {
 				t.Fatal(err.Error())
 			}
@@ -169,8 +230,8 @@ func TestIsConditions(t *testing.T) {
 	for _, tt := range []struct {
 		name       string
 		conditions []operatorv1.OperatorCondition
-		t          arov1alpha1.CondititionType
-		f          func([]operatorv1.OperatorCondition, arov1alpha1.CondititionType) bool
+		t          string
+		f          func([]operatorv1.OperatorCondition, string) bool
 		expect     bool
 	}{
 		{

```

**Review by m1kola — APPROVED** [review]
Looks good. I rebased becuase there were conflicts after merging https://github.com/Azure/ARO-RP/pull/1589. I hope it is ok.

I'll merge once e2es pass.

**Pull request merged**
